sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -237,23 +284,58 @@ batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
             case push span_ builder of
               Nothing -> (builder, True)
               Just b' -> (b', False)
-          when appendFailed $ void $ tryPutMVar workSignal Flush
-      , processorForceFlush = void $ tryPutMVar workSignal Flush
+          when appendFailed $ void $ atomically $ tryPutTMVar workSignal ()
+      , processorForceFlush = void $ atomically $ tryPutTMVar workSignal ()
       , -- TODO where to call restore, if anywhere?
-        processorShutdown = async $ mask $ \_restore -> do
-          -- flush remaining messages
-          void $ tryPutMVar workSignal Shutdown
-
-          shutdownResult <-
-            timeout (millisToMicros exportTimeoutMillis) $
-              wait worker
-          -- make sure the worker comes down
-          uninterruptibleCancel worker
-          -- TODO, not convinced we should shut down processor here
-
-          case shutdownResult of
-            Nothing -> pure ShutdownFailure
-            Just _ -> pure ShutdownSuccess
+        processorShutdown =
+          asyncWithUnmask $ \unmask -> unmask $ do
+            -- we call asyncWithUnmask here because the shutdown action is
+            -- likely to happen inside of a `finally` or `bracket`. the
+            -- @safe-exceptions@ pattern (followed by unliftio as well)
+            -- will use uninterruptibleMask in an exception cleanup. the
+            -- uninterruptibleMask state means that the `timeout` call
+            -- below will never exit, because `wait worker` will be in the
+            -- `uninterruptibleMasked` state, and the timeout async
+            -- exception will not be delivered.
+            --
+            -- see note [Unmasking Asyncs]
+            mask $ \_restore -> do
+              -- is it a little silly that we unmask and remask? seems
+              -- silly! but the `mask` here is doing an interruptible mask.
+              -- which means that async exceptions can still be delivered
+              -- if a process is blocking.
```

Comment:
question: given that we `unmask` here, there would be a race condition where someone could *theoretically* stuff in an inconveniently placed exception between the `unmask` and `mask` and stop our clean shutdown here, no?

I feel like we want to transition directly from `MaskedUninterruptible` -> `MaskedInterruptible`, rather than `MaskedUninterruptible` -> `Unmasked` -> `MaskedInterruptible`. https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Exception.html#t:MaskingState

Reading the docs of Control.Exception, I think that we want to use the function `interruptible` instead?

God, this is way too hard code; I really wish that this were abstracted into a library or something so it only has to be right once.

Reply:
You're right - we have a brief period of vulnerability to async exceptions.

```haskell
uninteruptibleMask_ $ 
  -- invincible
  asyncWithUnmask $ \unmask -> do
    -- still invincible
    unmask $ do
      -- totally vincible
      mask $ \restore -> do
        -- async exceptions only happen in blocking operations   
```

Fortunately, this period of vulnerability is very short, and I think it may not even be a problem. Async exceptions are only *really* visible at "yield points" (not a technical term) - with allocation being a yield point. `unsafeUnmask`, the [function that `unmask` here refers to](https://hackage.haskell.org/package/base-4.17.0.0/docs/src/GHC.IO.html#unsafeUnmask), unwraps the `IO` newtype and calls a primitive on it. `mask` might allocate if `getMaskingState` allocates, but I'd suspect that the allocation can be erased since we're immediately pattern matching on the result.

So uhh the safety here is *kinda* dependent on optimizations? Oof. Not sure we can really improve the situation. Especially since the only way we'll get this if we do somethin like

```haskell
cancel =<< shutdownProcessor processor
```

which, well, what do you expect from that?

`interruptible` is *almost* right: 

> When called outside [mask](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Exception.html#v:mask), or inside [uninterruptibleMask](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Exception.html#v:uninterruptibleMask), this function has no effect.

The problem is that we specifically need it if we're in `uninterruptibleMask`. It does seem like we're missing a primitive that does the transform you're talking about.

Reply:
Follow up GHC issue: https://gitlab.haskell.org/ghc/ghc/-/issues/22389